### PR TITLE
Rewrite X86 string instructions as intrinsics (strlen, memcpy, memcmp)

### PR DESCRIPTION
**Pull request by ptomin**
Intrisics is better than loops.
Currently 3 cases implemented:
```
rep movs -> memcpy
repe cmps -> memcmp
repne scasb (eax=0,ecx=0xFFFFFFFF) -> strlen
```
Segmented mode is not supported.

### CODE REVIEW

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            case Mnemonic.movsb:
+                // rep movs
+                if (IsRep())
+                    return RewriteMovsIntrinsic();
+                break;
+            case Mnemonic.cmps:
+            case Mnemonic.cmpsb:
+                // repe cmps
+                if (IsRepE())
+                    return RewriteCmpsIntrinsic();
```
Comment (uxmal) [inline src/Arch/X86/X86Rewriter.StringIntrinsics.cs:62]: ...and `RewriteCmpsToFindFirstDifference`

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+using Reko.Core.Types;
+
+namespace Reko.Arch.X86
+{
+    /// <summary>
+    /// Rewrite string intrinsics(strlen, memcpy, etc.).
+    /// </summary>
+    public partial class X86Rewriter
+    {
+        private static readonly IntrinsicProcedure findFirstChange = new IntrinsicBuilder("__find_first_change__", false)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/X86Rewriter.StringIntrinsics.cs]: Done in 6391592

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            if (arch.ProcessorMode.PointerType == PrimitiveType.SegPtr32)
+                return false;
+            if (!IsAutoIncrementingMode())
+                return false;
+            switch (instrCur.Mnemonic)
+            {
+            case Mnemonic.scasb:
+                // repne scasb
+                if (IsRepNe())
+                    return RewriteScasbIntrinsic();
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/X86Rewriter.StringIntrinsics.cs]: Done in d2c0582

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                break;
+            case Mnemonic.movs:
+            case Mnemonic.movsb:
+                // rep movs
+                if (IsRep())
+                    return RewriteMovsIntrinsic();
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/X86Rewriter.StringIntrinsics.cs]: Done in d2c0582

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs (hunk 1 above)
Comment (ptomin) [reply src/Arch/X86/X86Rewriter.StringIntrinsics.cs]: Done in d2c0582

File: src/Core/Intrinsics/CommonOps.cs [hunk 5]
```diff
@@ -136,6 +136,20 @@ public static class CommonOps
             .Param("T")
             .Void();
 
+        public static readonly IntrinsicProcedure Strlen = new IntrinsicBuilder("strlen", true)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Core/Intrinsics/CommonOps.cs:139]: I have replaced Int32 with UInt32 in f499a18. Should general solution be added to this PR?

File: src/Core/IntrinsicProcedure.cs [hunk 6]
```diff
@@ -169,7 +169,53 @@ protected virtual IntrinsicProcedure DoMakeInstance(DataType[] concreteTypes, Fu
         public IntrinsicProcedure MakeInstance(params DataType[] concreteTypes)
             => MakeInstance(0, concreteTypes);
 
-		public override string ToString()
+        public IntrinsicProcedure ResolvePointers(int ptrSize)
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Core/IntrinsicProcedure.cs:181]: Done in c52f727

File: src/Arch/X86/X86Rewriter.StringIntrinsics.cs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            m.Assign(
+                binder.EnsureFlagGroup(X86Instruction.DefCc(Mnemonic.cmp)!),
+                m.Cond(m.Const(instrCur.dataWidth, 0)));
+            return true;
+        }
+
+        private bool RewriteMovsIntrinsic()
+        {
+            var regCx = orw.AluRegister(Registers.rcx, instrCur.addrWidth);
+            var size = binder.CreateTemporary("size", instrCur.addrWidth);
```
[the remark this answers is not in the document]
Comment (ptomin) [reply src/Arch/X86/X86Rewriter.StringIntrinsics.cs:90]: Done in 657c187.

**Comment by uxmal** [comment]
Looks good! Thanks for the PR!

**Pull request merged**